src/System.Collections.Immutable/tests/ImmutableListBuilderTest.cs

```diff
@@ -178,6 +178,17 @@ public void Remove()
         }
 
         [Fact]
+        public void RemoveAllBugTest()
+        {
+            var builder = ImmutableList.CreateBuilder<int>();
+            var elemsToRemove = new[]{0, 1, 2, 3, 4, 5}.ToImmutableHashSet();
+            foreach(var elem in new[]{0, 1, 2, 3, 4, 5, 6})
+                builder.Add(elem);
```

Comment:
My best suspicion is that there is more than one 'balanced' tree given the number of elements you have. Perhaps with a different data set, you'd find one that fails for `AddRange`. Perhaps a larger collection?